File: Javascript/103. Binary Tree Zigzag Level Order Traversal.js
/*
Given the root of a binary tree, imagine yourself standing on the right side of it, 
    return the values of the nodes you can see ordered from top to bottom.
*/
import { buildTree, TreeNode } from "./buildTree.js";    
/**
 * 103. Binary Tree Zigzag Level Order Traversal
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if (root === null) { return []; }
    const result = [];
    const queue = [root];
    let zigzag = false;
    
    while (queue.length > 0) {
        const levelSize = queue.length;
        const levelNodes = [];
        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift();
            levelNodes.push(node.val); 
            if (node.left !== null) { queue.push(node.left); }
            if (node.right !== null) { queue.push(node.right); }            
        }
        if (zigzag) {
            levelNodes.reverse();
        }   
        result.push(levelNodes);
        zigzag = !zigzag;
    }
    return result;
};
//const root = buildTree([3,9,20,null,null,15,7]); // Expect [[3],[20,9],[15,7]]
//const root = buildTree([1]); // Expect [[1]]
const root = buildTree([]); // Expect []
console.log(zigzagLevelOrder(root));

File: Python/103. Binary Tree Zigzag Level Order Traversal.py
"""
Given the root of a binary tree, imagine yourself standing on the right side of it, 
    return the values of the nodes you can see ordered from top to bottom.
"""
# 103. Binary Tree Zigzag Level Order Traversal
from typing import Optional, List
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        
        result = []
        queue = deque([root])
        zigzag = False
        
        while len(queue) > 0:
            levelSize = len(queue)
            levelNodes = deque()
            for _ in range(0, levelSize):
                node = queue.popleft()
                if zigzag:
                    levelNodes.appendleft(node.val)
                else:
                    levelNodes.append(node.val)
                    
                if node.left is not None:
                    queue.append(node.left)
                    
                if node.right is not None:
                    queue.append(node.right)
            
            zigzag = not zigzag
            result.append(list(levelNodes))        
                
        return result
    
# Test Code
solution = Solution()
root = TreeNode(3, 
    TreeNode(9, None, None),
    TreeNode(20,
        TreeNode(15, None, None),
        TreeNode(7, None, None)
    )
) # Expect [[3],[20,9],[15,7]]

print(solution.zigzagLevelOrder(root))